### PR TITLE
Out-of-service form can be saved without a service being filled as provided (offered)

### PR DESCRIPTION
**Pull request by ekigamba**

### CODE REVIEW

File: opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java [hunk 1]
```diff
@@ -334,13 +333,13 @@ private void stockWastedVialsEnteredinIssuedForm(String key, String value) {
     private void stockDateEnteredinReceivedForm(String key, String value) {
         JSONObject object = getStep("step1");
         try {
-            if (object.getString("title").contains("Stock Received")) {
+            if (object.getString(getString(R.string.title_key)).contains(STOCK_RECEIVED)) {
                 if (key.equalsIgnoreCase("Date_Stock_Received") && value != null && !value.equalsIgnoreCase("")) {
 //                    if(balancetextview == null) {
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java]: can we delete all this commented code?

File: opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java [hunk 2]
```diff
@@ -401,8 +400,8 @@ private void stockDateEnteredinReceivedForm(String key, String value) {
     private void stockVialsenteredinReceivedForm(String key, String value) {
         JSONObject object = getStep("step1");
         try {
-            if (object.getString("title").contains("Stock Received")) {
-                if (key.equalsIgnoreCase("Vials_Received") && value != null && !value.equalsIgnoreCase("")) {
+            if (object.getString(getString(R.string.title_key)).contains(STOCK_RECEIVED)) {
+                if (key.equalsIgnoreCase(getString(R.string.vials_received_key)) && value != null && !value.equalsIgnoreCase("")) {
 //                    if(balancetextview == null) {
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java]: this too

File: opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java [hunk 3]
```diff
@@ -467,13 +466,13 @@ private void stockVialsenteredinReceivedForm(String key, String value) {
     private void stockDateEnteredinAdjustmentForm(String key, String value) {
         JSONObject object = getStep("step1");
         try {
-            if (object.getString("title").contains("Stock Loss/Adjustment")) {
+            if (object.getString(getString(R.string.title_key)).contains(STOCK_LOSS_ADJUSTMENT)) {
                 if (key.equalsIgnoreCase("Date_Stock_loss_adjustment") && value != null && !value.equalsIgnoreCase("")) {
 //                    if(balancetextview == null) {
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java]: this too

File: opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java [hunk 4]
```diff
@@ -533,8 +532,8 @@ private void stockDateEnteredinAdjustmentForm(String key, String value) {
     private void stockVialsenteredinAdjustmentForm(String key, String value) {
         JSONObject object = getStep("step1");
         try {
-            if (object.getString("title").contains("Stock Loss/Adjustment")) {
-                if (key.equalsIgnoreCase("Vials_Adjustment") && value != null && !value.equalsIgnoreCase("")) {
+            if (object.getString(getString(R.string.title_key)).contains(STOCK_LOSS_ADJUSTMENT)) {
+                if (key.equalsIgnoreCase(getString(R.string.vials_adjustment_key)) && value != null && !value.equalsIgnoreCase("")) {
 //                    if(balancetextview == null) {
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java]: this too

File: opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java [hunk 5]
```diff
@@ -610,5 +609,35 @@ public boolean checkIfBalanceNegative() {
...
+                            }
+                        }
+                    } else if (vaccineGroup.has(getString(R.string.key)) && vaccineGroup.getString(getString(R.string.key)).equals("Weight_Kg")
+                            && vaccineGroup.has(getString(R.string.value_key)) && vaccineGroup.getString(getString(R.string.value_key)).length() > 0) {
+                        return true;
+                    }
+                }
+            }
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/activity/PathJsonFormActivity.java]: is there a better way to log this?

File: opensrp-path/src/main/java/org/smartregister/path/fragment/PathJsonFormFragment.java [hunk 6]
```diff
@@ -408,23 +408,35 @@ public String getRelevantTextViewString(String currentKey) {
 
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
-        boolean balancecheck = true;
+        boolean balancecheck = true,
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/fragment/PathJsonFormFragment.java]: you could update this var to put it in proper camelCase, `balanceCheck`

File: opensrp-path/src/main/java/org/smartregister/path/fragment/PathJsonFormFragment.java [hunk 7]
```diff
@@ -408,23 +408,35 @@ public String getRelevantTextViewString(String currentKey) {
...
+                }
             } catch (Exception e) {
                 Log.e(getClass().getName(), e.toString(), e);
             }
         }
-        if (balancecheck) {
+        if (balancecheck && fillFormCheck) {
             return super.onOptionsItemSelected(item);
         } else {
+            String errorMessage = "";
```
Comment (pld) [inline opensrp-path/src/main/java/org/smartregister/path/fragment/PathJsonFormFragment.java]: what about DRYing this up a bit:

```java
String errorMessage = getString(balancecheck ? R.string.fill_form_check_error_msg : R.string.balance_error_msg);
```

Comment (ekigamba) [reply opensrp-path/src/main/java/org/smartregister/path/fragment/PathJsonFormFragment.java]: Cool

**Pull request closed without merging**